### PR TITLE
nixos/nextcloud: Add services.nextcloud.autoUpdateApps

### PR DESCRIPTION
**Pull request by maralorn**
###### Motivation for this change

This change adds the options `services.nextcloud.autoUpdateApps.enable` and `services.nextcloud.autoUpdateApps.startAt`.

If set it configures a systemd-timer unit to run `occ app:update --all`at the specified time.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by flokli** [comment]
@maralorn Thanks for the PR! Could you add a paragraph to `nixos/modules/services/web-apps/nextcloud.xml` explaining apps are currently installed imperatively, and link to the options introduced in this PR?

Maybe also enable it in the tests, so we hit the code paths there (even though we obviously can't test timers get fired)

**Comment by maralorn** [comment]
@flokli There you go!

File: nixos/modules/services/web-apps/nextcloud.xml [hunk 1]
```diff
@@ -111,5 +111,11 @@
    <link xlink:href="https://github.com/NixOS/nixpkgs/issues/49783">#49783</link>,
    for now it's unfortunately necessary to manually work around these issues.
   </para>
+
+  <para>
+   Right now app installation and configuration is done imperatively in the nextcloud web ui or via the <literal>nextcloud-occ</literal> comandline utility.
```
Comment (flokli) [inline nixos/modules/services/web-apps/nextcloud.xml]:
```suggestion
   Right now app installation and configuration is done imperatively in the nextcloud web ui or via the <literal>nextcloud-occ</literal> command line utility.
```

File: nixos/modules/services/web-apps/nextcloud.nix [hunk 2]
```diff
@@ -257,6 +257,23 @@ in {
...
+        description = ''
+          Run a auto update of all installed apps from the nextcloud repository.
+        '';
+      };
+      startAt = mkOption {
+        type = with types; either str (listOf str);
+        default = "05:00:00";
+        example = "Sun 14:00:00";
+        description = ''
+          When to run the update. See `systemd.services.<name>.startAt`.
```
Comment (flokli) [inline nixos/modules/services/web-apps/nextcloud.nix]: Seems the `<name>` doesn't get escaped when generating manpages:

```
⇒  nix-build nixos/release.nix -A manpages.x86_64-linux
these derivations will be built:
  /nix/store/mdc1snp1km5y963xc8xw8s5x6w39c3iq-options-db.xml.drv
  /nix/store/6p294xw0frkajlj27dk20wa732pyx4ps-generated-docbook.drv
  /nix/store/729zam31kvlb7b9267mfbrqm99df0vr9-nixos-manual-combined.drv
  /nix/store/xl8vg8w0ykf4g7pnkpsk6wpy4iqksfmg-manual-olinkdb.drv
  /nix/store/w2y87icgfnykcxly87kyxfxcpkp0izj9-nixos-manpages.drv
building '/nix/store/mdc1snp1km5y963xc8xw8s5x6w39c3iq-options-db.xml.drv'...
intermediate.xml:30885: parser error : Opening and ending tag mismatch: name line 30884 and para
</para></nixos:option-description><para><emphasis>Type:</emphasis> string or lis
       ^
intermediate.xml:30885: parser error : Opening and ending tag mismatch: para line 30884 and nixos:option-description
</para></nixos:option-description><para><emphasis>Type:</emphasis> string or lis
                                  ^
intermediate.xml:30891: parser error : Opening and ending tag mismatch: option-description line 30884 and listitem
            </filename></member></simplelist></listitem></varlistentry><varliste
                                                        ^
intermediate.xml:30891: parser error : Opening and ending tag mismatch: listitem line 30884 and varlistentry
            </filename></member></simplelist></listitem></varlistentry><varliste
                                                                       ^
intermediate.xml:59111: parser error : Opening and ending tag mismatch: varlistentry line 30884 and variablelist
      </filename></member></simplelist></listitem></varlistentry></variablelist>
                                                                               ^
intermediate.xml:59111: parser error : Opening and ending tag mismatch: variablelist line 3 and appendix
lename></member></simplelist></listitem></varlistentry></variablelist></appendix
                                                                               ^
intermediate.xml:59113: parser error : Premature end of data in tag appendix line 3

^
unable to parse intermediate.xml
note: keeping build directory '/tmp/nix-build-options-db.xml.drv-1'
builder for '/nix/store/mdc1snp1km5y963xc8xw8s5x6w39c3iq-options-db.xml.drv' failed with exit code 6
cannot build derivation '/nix/store/6p294xw0frkajlj27dk20wa732pyx4ps-generated-docbook.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/729zam31kvlb7b9267mfbrqm99df0vr9-nixos-manual-combined.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/w2y87icgfnykcxly87kyxfxcpkp0izj9-nixos-manpages.drv': 1 dependencies couldn't be built
error: build of '/nix/store/w2y87icgfnykcxly87kyxfxcpkp0izj9-nixos-manpages.drv' failed
```

Looking at intermediate.xml:

```
            </filename></member></simplelist></listitem></varlistentry><varlistentry><term xlink:href="#opt-services.nextcloud.autoUpdateApps.startAt" xml:id="opt-services.nextcloud.autoUpdateApps.startAt"><option>services.nextcloud.autoUpdateApps.startAt</option></term><listitem><nixos:option-description><para>When to run the update. See `systemd.services.<name>.startAt`.
```

Docbook inserts an unescaped `<name>` tag

Can you change this to ` When to run the update. See `systemd.services.&lt;name&gt;.startAt`.` to workaround that for now?

**Review by flokli — CHANGES_REQUESTED** [review]
Thanks for the followup!

File: nixos/modules/services/web-apps/nextcloud.nix (hunk 2 above)
Comment (flokli) [reply nixos/modules/services/web-apps/nextcloud.nix]: For completeness, I opened https://github.com/NixOS/nixpkgs/issues/61800, but I guess for now, escaping on your side for now should be fine :-)

**Review by flokli — APPROVED** [review]
LGMT :-)

Can you squash this together?

File: nixos/modules/services/web-apps/nextcloud.xml [hunk 3]
```diff
@@ -111,5 +111,11 @@
    <link xlink:href="https://github.com/NixOS/nixpkgs/issues/49783">#49783</link>,
    for now it's unfortunately necessary to manually work around these issues.
   </para>
+
+  <para>
+   Right now app installation and configuration is done imperatively in the nextcloud web ui or via the <literal>nextcloud-occ</literal> command line utility.
+   You can activate auto updates for your apps via
+   <literal><link linkend="opt-services.nextcloud.autoUpdateApps">services.nextcloud.autoUpdateApps</link></literal>.
```
Comment (flokli) [inline nixos/modules/services/web-apps/nextcloud.xml]:
```suggestion
   <literal><link linkend="opt-services.nextcloud.autoUpdateApps.enable">services.nextcloud.autoUpdateApps</link></literal>.
```

**Review by flokli — CHANGES_REQUESTED** [review]
In fact, I missed that part - references only exist on leaf nixos options, not all nodes. Sorry for the noise!

**Review by flokli — APPROVED** [review]

**Comment by flokli** [comment]
Thanks a lot!

**Comment by maralorn** [comment]
Can we backport this to 19.03? It's just a new option. That shouldn't break anything.

**Comment by flokli** [comment]
backported to 19.03 in 246ebba4f341c9e21fd591872dd29de18865f5cc.

**Pull request merged**